fx.go

```diff
@@ -19,3 +19,149 @@
 // THE SOFTWARE.
 
 package fx // import "go.uber.org/fx"
+
+import (
+	"io"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"github.com/uber-go/tally"
+
+	"go.uber.org/dig"
+	"go.uber.org/fx/config"
+	"go.uber.org/fx/metrics"
+	"go.uber.org/fx/ulog"
+	"go.uber.org/zap"
+)
+
+// Component is a function that initializes and returns objects to be shared across the code
+type Component interface{}
+
+// Module is a building block or primary function performed by a service built on fx
+type Module interface {
+	Name() string
+	Constructor() []Component
```

Comment:
Just remember a piece of feedback I forgot to include earlier.

I think it's critical for `Constructor()` to take in `config.Provider` because it will allow modules to consider config before returning components. For example, maybe module can always return components A and B, but an optional C needs to be enabled through config.